DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/data_bundle/src/data_bundle.cpp

```diff
@@ -0,0 +1,187 @@
+#include <iomanip>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+#include "memory_utils.hpp"
+#include "metaprogramming_utils.hpp"
+#include "unrolled_loop.hpp"
+
+using namespace sycl;
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename T, int n>
+struct DataBundle {
+  static constexpr size_t size = n;
+  using value_type = T;
+
+  DataBundle() {}
+
+  DataBundle(const T op) {
+    #pragma unroll
+    for (int idx = 0; idx < n; idx++) {
+      data_[idx] = op;
+    }
+  }
+
+  DataBundle(const DataBundle &op) {
+    #pragma unroll
+    for (int idx = 0; idx < n; idx++) {
+      data_[idx] = op.data_[idx];
+    }
+  }
+
+  DataBundle &operator=(const DataBundle &op) {
+    #pragma unroll
+    for (int idx = 0; idx < n; idx++) {
+      data_[idx] = op.data_[idx];
+    }
+    return *this;
+  }
+
+  // get a specific value in the bundle
+  T& operator[](int i) { return data_[i]; }
+  const T& operator[](int i) const { return data_[i]; }
+
+  // get a raw pointer to underlying data
+  T *Data() { return &data_[0]; }
+
+private:
+  T data_[n];  
+};
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ELEMENTS_PER_CYCLE
+#define ELEMENTS_PER_CYCLE 1
+#endif
+constexpr int kElementsPerCycle = ELEMENTS_PER_CYCLE;
+static_assert(kElementsPerCycle > 0);
+using PipeType = DataBundle<int, kElementsPerCycle>;
+
+using BufferPipe = ext::intel::pipe<class BufferPipeID, PipeType>;
+using USMPipe = ext::intel::pipe<class USMPipeID, PipeType>;
+
+class BufferProducerID;
+class BufferConsumerID;
+class USMProducerID;
+class USMConsumerID;
+
+event SubmitBufferProducer(queue &q, buffer<int, 1> &in_buf, size_t count) {
```

Comment:
Should we also encapsulate these producers/consumers into a shared header?  Don't you already have this in the FakeIOPipes?

Reply:
I was trying to avoid having anything about SYCL queues in the shared header files.

Reply:
Do you think that's an important goal?  I think if it's generically useful, which this clearly is, we should try to put it into the include directory.  Maybe we isolate it into a special sub-directory?

Reply:
I see them as orthogonal, but I understand the confusion because of what I named this project.

I think we should have a generic "bundle" class that has operator[] semantics, but that didn't end up being my goal here (despite the name, sorry).

The goal here is to make the MemoryToPipe and PipeToMemory, and don't think these methods should depend on a _specific_ bundle type.

Reply:
Wow. Early monday. My response makes no sense.

I have no problem throwing the actual kernels into the file, but I kind of like the user having to wrap all of our utilities in kernels. That way they know what they are submitting.